File: framework/python/src/soda/unittest/job.py
import json
import sys
import time

from soda.unittest.codec import CodecFactory

class JobEntry:

    @classmethod
    def run(cls, jobclass, method, ret_type, arg_types):
        cls.runJob(jobclass, method, ret_type, arg_types, validate_by_object=False)

    @classmethod
    def runWithObjectCheck(cls, jobclass, method, ret_type, arg_types, object_validator):
        cls.runJob(
                jobclass, method, ret_type, arg_types,
                validate_by_object=True,
                object_validator=object_validator)

    @classmethod
    def runWithSerialCheck(cls, jobclass, method, ret_type, arg_types, serial_validator):
        cls.runJob(
                jobclass, method, ret_type, arg_types,
                validate_by_object=False,
                serial_validator=serial_validator)

    @staticmethod
    def runJob(jobclass, method: str, ret_type, arg_types, *, 
            validate_by_object=False, 
            object_validator=None, 
            serial_validator=None):
        jspec = JobSpec()
        jspec.cls = jobclass
        jspec.method = method
        jspec.arg_types = arg_types
        jspec.ret_type = ret_type
        jspec.validate_by_object = validate_by_object
        jspec.object_validator = object_validator
        jspec.serial_validator = serial_validator
        
        runner = JobRunner()
        runner.run(jspec)

class InputData:

    def __init__(self, obj):
        self.obj = obj

    @property
    def expected(self):
        return self.obj['expected']

    @property
    def id(self):
        return self.obj['id']

    @property
    def args(self):
        return self.obj['args'][:]

    def arg(self, index):
        return self.obj['args'][index]

class JobSpec:
    pass

class JobRunner:

    def run(self, jbs: JobSpec) -> None:
        input_data = InputData(json.load(sys.stdin))
        args = (CodecFactory.create(t).decode(a) for t, a in zip(jbs.arg_types, input_data.args))

        method = getattr(jbs.cls, jbs.method)
        obj = jbs.cls()

        start = time.time()
        res = method(obj, *args)
        end = time.time()

        serial_res = CodecFactory.create(jbs.ret_type).encode(res)
        elapseMillis = (end - start) * 1000

        resp = {
            'id': input_data.id,
            'result': serial_res,
            'elapse': elapseMillis
        }

        success = True
        if input_data.expected is not None:
            if jbs.validate_by_object:
                vf = jbs.object_validator or (lambda x, y: x == y)
                expc = CodecFactory.create(jbs.ret_type).decode(input_data.expected)
                success = vf(expc, res)
            else:
                vf = jbs.serial_validator or (lambda x, y: x == y)
                success = vf(input_data.expected, serial_res)

        resp['success'] = success

        print(json.dumps(resp))

File: framework/ruby/src/soda/unittest/job.rb
require 'json'
require 'time'

require 'soda/unittest/codec'

module Soda end

module Soda::Unittest

class JobEntry
    def self.run(function, ret_type, arg_types)
        runJob(function, ret_type, arg_types, false, nil, nil)
    end

    def self.runWithObjectCheck(function, ret_type, arg_types, object_validator)
        runJob(function, ret_type, arg_types, true, object_validator, nil)
    end

    def self.runWithSerialCheck(function, ret_type, arg_types, serial_validator)
        runJob(function, ret_type, arg_types, false, nil, serial_validator)
    end

    def self.runJob(function, ret_type, arg_types, validate_by_object=false, object_validator=nil, serial_validator=nil)
        jspec = JobSpec.new(function, ret_type, arg_types)
        jspec.validate_by_object = validate_by_object
        jspec.object_validator = object_validator
        jspec.serial_validator = serial_validator
        JobRunner.new.run(jspec)
    end
end

class InputData
    def initialize(_js)
        @json = _js
    end

    def expected
        @json['expected']
    end

    def id
        @json['id']
    end

    def args
        @json['args']
    end

    def arg(index)
        @json['args'][index]
    end
end

class JobSpec
    def initialize(function, ret_type, arg_types)
        @function = function
        @ret_type = ret_type
        @arg_types = arg_types
        @validate_by_object = false
        @object_validator = nil
        @serial_validator = nil
    end

    attr_accessor :function, :ret_type, :arg_types, :validate_by_object, :object_validator, :serial_validator
end

class JobRunner
    def run(jspec)
        input_data = InputData.new(JSON.parse(ARGF.read))
        args = jspec.arg_types.each_with_index.map { |t,i|
            decode(t, input_data.arg(i))
        }

        start_time = Time.now
        res = jspec.function.call(*args)
        end_time = Time.now

        serial_res = encode(jspec.ret_type, res)
        elapse_ms = (end_time - start_time) * 1000.0

        resp = {
            'id' => input_data.id,
            'result' => serial_res,
            'elapse' => elapse_ms
        }

        success = true
        if !input_data.expected.nil?
            if jspec.validate_by_object
                expc = decode(jspec.ret_type, input_data.expected)
                vf = jspec.object_validator || -> (e, r) { e == r }
                success = vf.call(expc, res)
            else
                vf = jspec.serial_validator || -> (e, r) { e == r }
                success = vf.call(input_data.expected, serial_res)
            end
        end

        resp['success'] = success

        puts resp.to_json
    end

    def encode(obj_type, obj)
        CodecFactory.create(obj_type).encode(obj)
    end

    def decode(obj_type, ser)
        CodecFactory.create(obj_type).decode(ser)
    end
end

end  # module Soda::Unittest
